HPCC-12750 Generate environment.xml during package install and configure

xwang2713 opened the pull request:
Call envgen to generate environment.xml to replace default environment.xml
in the repository. It will include changes of each component under <HPCC_INSTALL>/
componentfiles/configxml/ directory

Signed-off-by: Xiaoming Wang xiaoming.wang@lexisnexi.com

@garonsky please review

xwang2713 committed:
HPCC-12750 Generate environment.xml during package install and configure

Call envgen to generate environment.xml to replace default environment.xml
in the repository. It will include changes of each component under <HPCC_INSTALL>/
componentfiles/configxml/ directory

Remove default environment.xml and INSTALL code in cmake configuration files
Remove generated environmen.xml under install directory

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: Fail

HPCC Stop: OK
HPCC Uninstall: OK

mckellyln commented:
Should perhaps remove environment.xml.in as part of the PR so that is is clear new changes to environment at install need to come from component xsd.
What about testing environment.xml.in also ?
@AttilaVamos including you fyi.

xwang2713 commented:
Updated pull-request: remove default environment.xml.in and INSTALL code

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: Fail

HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
@wangkx @mckellyln The error message was: "hpcc-init: unrecognized service". Is the hpcc start method changed?

mckellyln commented:
I do not think this PR changes any service, just the environment.xml file that gets put into /etc/HPCCSystems at install time ?

xwang2713 commented:
That's right

AttilaVamos commented:
OK, I will check what's going on in smoketest for this pr.

garonsky commented:
I have a bug in the envgen that maybe causing this failure on startup.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: Fail
hpcc-init: unrecognized service

HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
I scheduled the last test to check the improvement of smoketest error handling, only.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: Fail
HPCC start:
hpcc-init: unrecognized service
HPCC status:
hpcc-init: unrecognized service

HPCC Stop: OK
HPCC Uninstall: OK

mckellyln commented:
How does this affect the existing environment.xml during an upgrade ?
Has it been tested ?  @Michael-Gardner fyi.

Michael-Gardner commented on initfiles/bash/etc/init.d/install-init.in:
```diff
@@ -182,6 +182,12 @@ homePath=$(cat /etc/passwd | grep -e "^${user}:" | cut -d':' -f6)
 
+# Generate environment.xml
+echo "Generate environment.xml"
+${path}/sbin/envgen -env ${path}/etc/HPCCSystems/rpmnew/environment.xml -ip . -roxienodes 1 -thornodes 1 -slavesPerNode 1 -o data4=/var/lib/[NAME]/hpcc-data4/[COMPONENT] -set_xpath_attrib_value Software/EclCCServerProcess/Option @name maxCompileThreads -set_xpath_attrib_value Software/EclCCServerProcess/Option @value 2 -override roxie,@allFilesDynamic,true -override roxie,@laxyOpen,smart -overrid ...
```
The following exception is causing the installation to fail.

Exception: -1:
Config file does not define values for Software/EspProcess/EspService[@name='WsDeploy_wsdeploy_esp']/LocalConfFile and Software/EspProcess/EspService[@name='WsDeploy_wsdeploy_esp']/LocalEnvConfFile

garonsky replied:
Yes, there is a bug I have to fix.

garonsky replied:
I double checked the parameters to envgen it doesn't appear that is the problem.

garonsky replied:
@xwang2713 indicated that updates will be unchanged in the logic.

Michael-Gardner replied:
@xwang2713 The envgen call is failing because it's failing to open a file at /etc/HPCCSystems/environment.conf.  We need to make sure that path has been laid down before we do the envgen call.

xwang2713 committed:
lay down environment.cfg before calling envgen
Move envgen call to a new file: create-env-xml.sh
Remove duplication environment.xml for regression test

fix typo "laxyOpen". It should be "lazyOpen".

xwang2713 commented:
Copy environment.cfg to /etc/HPCCSystems/ before generating environment.xml

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

Michael-Gardner commented:
@xwang2713 Installed fine with the changes.  Looks good.

richardkchapman commented:
@Michael-Gardner Needs rebasing for a clean merge

Michael-Gardner commented:
@xwang2713 This needs a rebase

xwang2713 commented:
@Michael-Gardner rebased with candidate-6.0.0

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

Michael-Gardner commented:
@xwang2713 tested on centos 6 and ubuntu 14.  Looks good.

richardkchapman commented on initfiles/bash/etc/init.d/install-init.in:
```diff
@@ -184,9 +184,17 @@ echo ""
+echo "Generate environment.xml"
+# Generate environment.xml
+${path}/sbin/envgen -env ${path}/etc/HPCCSystems/rpmnew/environment.xml -ip . -roxienodes 1 -thornodes 1 -slavesPerNode 1 -o data4=/var/lib/[NAME]/hpcc-data4/[COMPONENT] -set_xpath_attrib_value Software/EclCCServerProcess/Option @name maxCompileThreads -set_xpath_attrib_value Software/EclCCServerProcess/Option @value 2 -override roxie,@allFilesDynamic,true -override roxie,@laxyOpen,smart -overrid ...
+cp ${path}/etc/HPCCSystems/rpmnew/environment.xml ${path}/testing/regress/environment.xml || exit 1
```
What's the point of having a separate environment for the regression suite if it's always going to be the same as the standard one? I thought we used to have a different one because the settings appropriate for the regression suite were different?

xwang2713 replied:
@AttilaVamos do you still need a separate environment.xml for regression? It is under <HPCC INSTALL DIR>/testing/regression/

AttilaVamos replied:
No. I don't need it. If I need some special settings (debug/log level change) in test environment I do it with OBT/smoketest script on fly.

richardkchapman commented on initfiles/bash/etc/init.d/install-init.in:
```diff
@@ -184,9 +184,17 @@ echo ""
 
+echo "Generate environment.xml"
+# Generate environment.xml
+${path}/sbin/envgen -env ${path}/etc/HPCCSystems/rpmnew/environment.xml -ip . -roxienodes 1 -thornodes 1 -slavesPerNode 1 -o data4=/var/lib/[NAME]/hpcc-data4/[COMPONENT] -set_xpath_attrib_value Software/EclCCServerProcess/Option @name maxCompileThreads -set_xpath_attrib_value Software/EclCCServerProcess/Option @value 2 -override roxie,@allFilesDynamic,true -override roxie,@laxyOpen,smart -overrid ...
```
Feels like we should be taking those overrides from a file rather than all on command line

xwang2713 commented:
Move evngen call to a new file sbin/create-env-xml.sh.
Remove duplicate environment.xml for regression test.
@Michael-Gardner please help to review again.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community-with-plugins_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

Michael-Gardner commented:
@xwang2713 tested on ubuntu 14 and centos 6 again.  All is well.

richardkchapman commented:
Needs rebasing for a clean merge

xwang2713 commented:
I rebased with candidate-6.0.0.
@garonsky ,  this fix is still targeted for HPCC 6.0.0, right?

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

garonsky commented:
Yes, I think this should be for 6.0.

richardkchapman commented:
Has someone confirmed that the output from envgen is the same as the old environment.xml.in??

xwang2713 commented:
I will compare it.

xwang2713 commented:
New environment.xml file 1183 lines. The old one has 1084 lines.
There are 23 lines differences. 25 lines in new file only and 15 lines in old file only.
HPCC single node starts OK with new environment.xml. I will attach both xml files to JIRA HPCC-12750.

xwang2713 commented:
@garonsky  do we have the schema for environment.xml?

garonsky commented:
The non-default values that were previously set by hard coding the environment.xml have been incorporated in the commit through the envgen parameters here: `envgen -env ${OUTPUT} -ip . -roxienodes 1 -thornodes 1 -slavesPerNode 1 -o data4=/var/lib/[NAME]/hpcc-data4/[COMPONENT] -set_xpath_attrib_value Software/EclCCServerProcess/Option @name maxCompileThreads -set_xpath_attrib_value Software/EclCCServerProcess/Option @value 2 -override roxie,@allFilesDynamic,true -override roxie,@laxyOpen,smart -override roxie,@localSlave,true -override roxie,@slaveConfig,simple`

The other differences in the files are primarily ordering, naming, and the _missing_ settings that are now being picked up.

richardkchapman commented:
Was laxyOpen a typo in your comment or in the code?

richardkchapman commented on initfiles/sbin/create-env-xml.sh.in:
```diff
@@ -0,0 +1,29 @@
+OUTPUT=${INSTALL_DIR}/etc/HPCCSystems/rpmnew/environment.xml
+[ -n "$1" ] && OUTPUT=$1
+
+${INSTALL_DIR}/sbin/envgen -env ${OUTPUT} -ip . -roxienodes 1 -thornodes 1 -slavesPerNode 1 -o data4=/var/lib/[NAME]/hpcc-data4/[COMPONENT] -set_xpath_attrib_value Software/EclCCServerProcess/Option @name maxCompileThreads -set_xpath_attrib_value Software/EclCCServerProcess/Option @value 2 -override roxie,@allFilesDynamic,true -override roxie,@laxyOpen,smart -override roxie,@localSlave,true -over ...
```
laxyOpen looks like a typo to me...

richardkchapman commented:
I attached a "smart diff" of the two files to the Jira. Most significant difference I spotted on a quick glance is the missing thor-roxie cluster in the topology.

xwang2713 commented:
@garonsky could you check "@laxyOpen" option in envgen?  I think I got it from you but maybe somewhere it got unexpected latters.
Here are whole command and options:
+${INSTALL_DIR}/sbin/envgen -env ${OUTPUT} -ip . -roxienodes 1 -thornodes 1 -slavesPerNode 1 -o data4=/var/lib/[NAME]/hpcc-data4/[COMPONENT] -set_xpath_attrib_value Software/EclCCServerProcess/Option @name maxCompileThreads -set_xpath_attrib_value Software/EclCCServerProcess/Option @value 2 -override roxie,@allFilesDynamic,true -override roxie,@laxyOpen,smart -override roxie,@localSlave,true -override roxie,@slaveConfig,simple

xwang2713 commented:
Gleb will add thor-roxie cluster setting and re-exam the environment.xml template. After it I will build and test it again.

xwang2713 committed:
HPCC-12750 Add thor_roxie setting in genenvrules.conf

Adding thor_roxie settings to generate environment.xml matched with
previous supplied default environment.xml.
update option for envgen to change thor_roxie prefix.

xwang2713 commented:
@garonsky please review the changes in genenvrules.conf.  I did a build and tested single node.

garonsky commented:
@xwang2713 please use the latest genenvrules.conf.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
Looking closer but the prefix for thor_roxie needs to be "thor"

garonsky commented:
@xwang2713 to rename the prefix to thor add the following parameters to the envgen commmand line: `-set_xpath_attrib_value Software/Topology/Cluster[@name="thor_roxie"] @prefix thor`

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

xwang2713 commented:
Add option: "-set_xpath_attrib_value Software/Topology/Cluster[@name=\"thor_roxie\"] @prefix thor" to envgen in create-env-xml.sh script.

Generated portion of environment.xml: <Cluster name="thor_roxie" prefix="thor">

xwang2713 commented:
Gleb will review it.

garonsky commented:
https://github.com/hpcc-systems/HPCC-Platform/pull/8199 needs to be pulled to generated the proper environment.xml.

The changes look good.
